Employee login

jkhabra opened the pull request:
Resolves issue #41 and  #42

jkhabra committed:
wip: shows only project section on employee login

jkhabra committed:
wip: add crate project form

jkhabra committed:
wip: add crate project form

jkhabra committed:
feat: hasura migrations for giving user access to projects

jkhabra committed:
fix: lint issues in add project form

bitspook commented on ui/src/containers/DashboardLayout/style.module.scss:
```diff
@@ -95,6 +95,27 @@
+    flex-direction: column;
+  }
+
+  .sub-option {
```
Let's get rid of `sub-` prefix.

bitspook replied:
Looks like this class isn't used anywhere. Let's remove it if that's the case.

bitspook commented on ui/src/containers/DashboardLayout/index.tsx:
```diff
@@ -14,7 +17,55 @@ interface LayoutProps {
+      Projects
+    </div>
+    <Link href="/projects">
+      <div className={c('sub-container')}>{projects.map((p) => dropdownOptions(p))}</div>
```
Please check if `.row` class we have for admin section can work here.

jkhabra committed:
Merge branch 'master' into feature/employee-login

jkhabra committed:
feat: fetch user's organization, projects and employees in me API

jkhabra committed:
feat: update redux state on the response of me API

- when get me API's data than add user in auth, organizations in organizations
state and so on.

jkhabra committed:
feat: show dynamic projects in sidebar

jkhabra committed:
feat: implement crate project API

bitspook commented on ui/src/pages/projects/add/index.tsx:
```diff
@@ -0,0 +1,188 @@
+
+  useEffect(() => {
+    dispatch(fetchOrganization());
+  }, []);
```
1. Admin should see full list of organizations in this form
2. User should see only current organization

jkhabra committed:
fix: user can only see current organization

- admin can see full list
- create user selectors for get getting user's current organization-id and projects

bitspook commented on ui/src/entities/User/index.ts:
```diff
@@ -0,0 +1,110 @@
+import 'cross-fetch/polyfill';
```
We don't need to do it in every entity. I think we should move this to a common place where it is guaranteed to be executed. e.g `pages/_app.tsx`.

bitspook commented on ui/src/entities/User/selectors.ts:
```diff
@@ -0,0 +1,22 @@
+import { EmployeeState } from 'src/duck/employee';
+import { ProjectState } from 'src/duck/project';
+
+export const getUsersOrganizationId = () => {
```
A user can have multiple organizations. I think we should rename this to `getActiveOrganization` or `getActiveUserOrganization`. What do you think?

bitspook commented on ui/src/entities/User/selectors.ts:
```diff
@@ -0,0 +1,22 @@
+  return emp?.organization_id ? emp.organization_id : null;
+};
+
+export const getUsersProjects = () => {
```
`Users` shouldn't be a plural here. This function (I suppose) returns active projects of current user. Projects are many, but user isn't.

bitspook replied:
Also, please add type signatures for both these functions.

bitspook commented on ui/src/pages/projects/add/index.tsx:
```diff
@@ -0,0 +1,195 @@
+  >((state) => state.organization);
+
+  useEffect(() => {
+    dispatch(fetchOrganization());
```
Should we avoid hitting this endpoint if user isn't admin? For non-admin user, the only organization that goes in the form is his current organization.

bitspook commented on ui/src/pages/projects/add/index.tsx:
```diff
@@ -0,0 +1,195 @@
+    return <span>loading...</span>;
+  }
+
+  if (user?.role.id === 'USER' && orgId) {
```
I recommend we create an enum for valid user roles. I recommend doing so in `entities/User.ts`.

bitspook commented on ui/src/pages/projects/add/index.tsx:
```diff
@@ -0,0 +1,195 @@
+    return <span>loading...</span>;
+  }
+
+  if (user?.role.id === 'USER' && orgId) {
```
We can eliminate the possibility of `user` not being present, hence get rid of `?` in `user?` by either forcefully casting it to `User`, or by checking for empty value of user in an if block at top.

bitspook replied:
That gives an idea; we can probably create a `currentUser` selector in `entities/User/selectors` which guarantees that it'll give us a user. We can rely on forced conversion here because we know that `WitheUser` component won't let a user enter this component without him being logged in.

bitspook commented on ui/src/pages/projects/add/index.tsx:
```diff
@@ -0,0 +1,195 @@
+    const org = organizations.find((o) => o.id === orgId);
+    userOrganizations = [org];
+  }
+  userOrganizations = organizations;
```
What are we trying to do here? We checked if user's role is `USER`, and if it is, we assigned a single organization to `userOrganizations`. And then we assigned all organizations we could fetch to the user anyway. So the user.role check above isn't really doing anything here.
